pkg/flow/internal/controller/loader.go

```diff
@@ -181,20 +181,27 @@ func (l *Loader) WriteBlocks(debugInfo bool) []*hclwrite.Block {
 	return blocks
 }
 
-// Reevaluate reevaluates the arguments for c and any component which directly
-// or indirectly depends on c.
+// EvaluateDependencies re-evaluates components which depend directly or
+// indirectly on c. EvaluateDependencies should be called whenever a component
+// updates its exports.
 //
 // The provided parentContext can be used to provide global variables and
 // functions to components. A child context will be constructed from the parent
 // to expose values of other components.
-func (l *Loader) Reevaluate(parentContext *hcl.EvalContext, c *ComponentNode) {
+func (l *Loader) EvaluateDependencies(parentContext *hcl.EvalContext, c *ComponentNode) {
 	l.mut.RLock()
 	defer l.mut.RUnlock()
 
 	// Make sure we're in-sync with the current exports of c.
 	l.cache.CacheExports(c.ID(), c.Exports())
 
 	_ = dag.WalkReverse(l.graph, []dag.Node{c}, func(n dag.Node) error {
+		if n == c {
```

Comment:
Pedantically is this true? Can a component reference itself?

Reply:
> Can a component reference itself?

No, but the starting nodes passed to WalkReverse will be part of what's walked, so we need to ignore the node.

Reply:
> Can a component reference itself?

(technically they can until #1687 is resolved, but it'll be disallowed in the future)